text/3487-feature-visibility.md

```diff
@@ -0,0 +1,165 @@
+easier to understand and lowers the chance of library users accidentally using
+unstable internal API.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+There will be a new flag allowed within `[features]`: `public`. This is boolean
+flag defaulting to `true` that indicates whether or not downstream crates should
+be allowed to use this feature.
+
+```toml
+[features]
+foo = { enables = [], public = false}
+```
+
+Attempting to use a private feature on a downstream crate will result in
+messages like the following:
+
+```
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
+
+The default `true` is not consistent with [`public_private_dependencies`] or
+Rust's `pub`, but is a reasonable default to be consistent with the current
+behavior. This means that either `feature = []` or
+`feature = { "enables" = [] }` will result in the same configuration.
+
+The name `public` was chosen in favor of `pub` to be consistent with the
+[`public_private_dependencies`] RFC, and to match the existing style of using
+non-truncated words as keys.
+
+In general, marking a feature `public = false` should make tooling treat the
+feature as non-public API. This is described as the following:
+
+- The feature is always usable within the same crate:
+  - Enabled by other features, e.g.
+    `foo = { enables = [some-private-feature] }`, is allowed
+  - Referenced in `[[bench]]` and `[[test]]` target `required-features`
+  - Using the feature on the command-line is allowed
+- Users may explicitly specifying the private features for their dependencies
+  on the command-line (e.g. `--features somecrate/private-feature`) which would
+  otherwise be forbidden
+- The feature should not be accepted by `cargo add --features`
+- The feature should not be reported from `cargo add`'s feature output report
+  - A future tool like `cargo info` shouldn't display information about these
+    features
+- Once `rustdoc` is able to consume feature metadata, `rustdoc` should not
+  document these features unless `--document-private-items` is specified
+
+Attempting to use a private feature in any of the forbidden cases should result
+in an error. Exact details of how features work will likely be refined during
+implementation and experimentation.
+
+This feature requires adjustments to the index for full support. This RFC
+proposes that it would be acceptable for the first implementation to simply
+strip private features from the manifest; this means that there will be no way
+to `cfg` based on these features.
```

Comment:
It's not clear what `cfg` means here. I assumed it is the `cfg` attribute and the `cfg` macro. Could you expand and link to their doc respectively?

Reply:
I kind of forgot the exact context here, this comes from Ed's comment at https://github.com/rust-lang/rfcs/pull/3416#issuecomment-1587839516 and the two following comments. Also https://github.com/rust-lang/rfcs/pull/3416#discussion_r1177020997.

Do you have any ideas for how to improve the wording?

Thank you for taking a look by the way

Reply:
If we strip the private features then when cargo does a build, it won't tell rustc about private features and if you do `#[cfg(feature = "private-foo")]` then it will never evaluate to `true`